app/services/fact_importer.rb

```diff
@@ -4,22 +4,28 @@ class FactImporter
   delegate :logger, :to => :Rails
   attr_reader :counters
 
+  def deprecation_warning
+    Foreman::Deprecation.deprecation_warning('2.0', 'Use Importer Registry to register or obtain facts for a plugin')
+  end
+
   def self.importer_for(type)
-    importers[type.to_s]
+    deprecation_warning
+    @importers.get(type)
   end
 
   def self.importers
-    @importers ||= {}.with_indifferent_access
+    deprecation_warning
+    @importers ||= Foreman::Plugin::ImporterRegistry.new
   end
 
   def self.register_fact_importer(key, klass, default = false)
-    importers.default = klass if default
-
-    importers[key.to_sym] = klass
+    deprecation_warning
+    @importers.register(key, klass, default)
```

Comment:
Foreman::Plugin.fact_registry.
```suggestion
    Foreman::Plugin.fact_registry.register(key, klass, default)
```